src/Ouroboros-Mini_Protocols-Chain_Sync.thy

```diff
@@ -0,0 +1,226 @@
+
+abbreviation (input) tip :: "chain \<Rightarrow> block" where
+  "tip \<C> \<equiv> \<C> ! (tip_index \<C>)"
+
+subsection \<open>Points\<close>
+
+type_synonym point = hash
+
+definition block_point :: "block \<Rightarrow> point" where
+  [simp]: "block_point \<equiv> hash"
+
+subsection \<open>Protocol Parameters\<close>
+
+datatype phase =
+  is_intersection_finding: IntersectionFinding |
+  is_chain_update: ChainUpdate
+
+type_synonym consumer_parameters = "
+  chain \<comment> \<open>current chain\<close> \<times>
+  phase \<comment> \<open>current phase\<close> \<times>
+  (chain \<Rightarrow> point list)" \<comment> \<open>function for selecting candidate points\<close>
+
+type_synonym producer_parameters = "
+  chain \<comment> \<open>current chain\<close> \<times>
+  chain_index \<comment> \<open>read pointer\<close> \<times>
+  bool \<comment> \<open>whether the next response to \<open>RequestNext\<close> should be \<open>RollBackward\<close>\<close>"
+
+locale chain_sync =
+  fixes initial_params\<^sub>c :: consumer_parameters
+    and initial_params\<^sub>p :: producer_parameters
+
+subsection \<open>Parties\<close>
+
+datatype party =
+  Producer |
+  Consumer
+
+subsection \<open>State Machines\<close>
+
+datatype state =
+  Idle |
+  Intersect |
+  CanAwait
+
+datatype message =
+  is_next_request: RequestNext |
+  is_roll_forward: RollForward \<open>block\<close> \<open>block\<close> |
+  is_roll_backward: RollBackward \<open>point\<close> \<open>block\<close> |
+  is_await_reply: AwaitReply |
+  is_find_intersect: FindIntersect \<open>point list\<close> |
+  is_intersect_found: IntersectFound \<open>point\<close> \<open>block\<close> |
+  is_intersect_not_found: IntersectNotFound \<open>block\<close>
+
+fun agent_in_state' where
+  "agent_in_state' Idle = Consumer" |
+  "agent_in_state' _ = Producer"
+
+inductive can_finish_in_state' where
+  "can_finish_in_state' Idle"
+
+declare can_finish_in_state'.simps [simp]
+
+primrec next_state' where
+  "next_state' Idle m = (
+    partial_case m of
+      FindIntersect _ \<Rightarrow> Intersect |
+      RequestNext \<Rightarrow> CanAwait)" |
+  "next_state' Intersect m = (
+    partial_case m of
+      IntersectFound _ _ \<Rightarrow> Idle |
+      IntersectNotFound _ \<Rightarrow> Idle)" |
+  "next_state' CanAwait m = (
+    partial_case m of
+      RollForward _ _ \<Rightarrow> Idle |
+      RollBackward _ _ \<Rightarrow> Idle |
+      AwaitReply \<Rightarrow> Idle)" \<comment> \<open>only for this initial implementation\<close>
+
+definition state_machine where
+  [simp]: "state_machine = \<lparr>
+    initial_state = Idle,
+    agent_in_state = agent_in_state',
+    can_finish_in_state = can_finish_in_state',
+    next_state = next_state'
+  \<rparr>"
+
+sublocale chain_sync \<subseteq> protocol_state_machine \<open>state_machine\<close> .
+
+subsection \<open>Programs\<close>
+
+definition rollback_to :: "chain \<Rightarrow> point \<Rightarrow> chain" where
```

Comment:
I guess we should rename this function to roll_back_to. The noun “rollback” is written without a space, but this function is about the action of rolling back, and “to roll back” is written with a space between “roll” and “back”.